### PR TITLE
feat: improve adhoc SQL validation

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

Improve the `validate_adhoc_subquery` function.

I changed the logic of `insert_rls`. As originally written we'd pass (1) the adhoc SQL, (2) the dataset name and (3) the associated RLS predicate. In order to use it here I modified it to receive (1) the adhoc SQL with the (2) database and (3) schema where it will run. We then parse the adhoc SQL, and for each table found we check if there's an associated RLS rule; if so, we modify the SQL to include the RLS predicate(s).

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

Updated unit tests, will add more.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -984,15 +992,14 @@ def is_alias_used_in_orderby(col: ColumnElement) -> bool:
 
     def _get_sqla_row_level_filters(
         self, template_processor: BaseTemplateProcessor
-    ) -> List[str]:
+    ) -> List[TextClause]:
```
Comment (betodealmeida) [inline superset/connectors/sqla/models.py:995]: Not sure why `mypy` didn't catch this.

File: superset/connectors/sqla/models.py [hunk 2]
```diff
@@ -1420,7 +1441,6 @@ def get_sqla_query(  # pylint: disable=too-many-arguments,too-many-locals,too-ma
                 and db_engine_spec.allows_hidden_cc_in_orderby
                 and col.name in [select_col.name for select_col in select_exprs]
             ):
-                validate_adhoc_subquery(str(col.expression))
```
Comment (betodealmeida) [inline superset/connectors/sqla/models.py:1423]: Moved this [up](https://github.com/apache/superset/pull/19454/files#diff-af5e15a55ff9c81f441eb6b9b10dae13b2ee17fec614d3e3ad0057ed0dc814aeR1155-R1160), where it's still a string.

File: tests/unit_tests/sql_parse_tests.py [hunk 3]
```diff
@@ -1391,13 +1392,37 @@ def test_has_table_query(sql: str, expected: bool) -> None:
...
     """
-    statement = sqlparse.parse(sql)[0]
     condition = sqlparse.parse(rls)[0]
-    assert str(insert_rls(statement, table, condition)).strip() == expected.strip()
+    add_table_name(condition, table)
+
+    # pylint: disable=unused-argument
+    def get_rls_for_table(
+        candidate: Token, database_id: int, default_schema: str
+    ) -> TokenList:
```
Comment (villebro) [inline tests/unit_tests/sql_parse_tests.py:1407]: nit:
```suggestion
    ) -> Optional[TokenList]:
```

**Review by villebro — COMMENTED** [review]
Tested to work very nicely except in one edge case (would be great to add a test for that case, too).

**Review by villebro — APPROVED** [review]

**Comment by sadpandajoe** [comment]
🏷️  preset:2022.11
